File: apps/api/internal/domain/logs/logs_entity.go
package logs

import (
	"encoding/json"
	"time"
)

// DeviceLog represents a device event log entry.
type DeviceLog struct {
	ID        string          `json:"id"`
	DeviceID  string          `json:"deviceId"`
	EventType string          `json:"eventType"`
	Timestamp time.Time       `json:"timestamp"`
	Data      json.RawMessage `json:"data,omitempty"`
}

// EventType constants for device logs.
const (
	EventTypeConnection     = "connection"
	EventTypeCommand        = "command"
	EventTypeTelemetry      = "telemetry"
	EventTypeError          = "error"
	EventTypeWarning        = "warning"
	EventTypeInfo           = "info"
	EventTypeUpdate         = "update"
	EventTypeRegistration   = "registration"
	EventTypeDeregistration = "deregistration"
)

// IsValidEventType checks if the event type is valid.
func IsValidEventType(eventType string) bool {
	switch eventType {
	case EventTypeConnection,
		EventTypeCommand,
		EventTypeTelemetry,
		EventTypeError,
		EventTypeWarning,
		EventTypeInfo,
		EventTypeUpdate,
		EventTypeRegistration,
		EventTypeDeregistration:
		return true
	default:
		return false
	}
}

File: packages/API_Client/src/domain/logs/logs-entity.ts
export type LogEventType =
  | "connection"
  | "command"
  | "telemetry"
  | "error"
  | "warning"
  | "info"
  | "update"
  | "registration"
  | "deregistration";

export interface LogEntry {
  id: string;
  deviceId: string;
  eventType: LogEventType;
  timestamp: Date;
  data?: Record<string, unknown>;
}

export interface LogListResult {
  logs: LogEntry[];
  hasMore: boolean;
  nextCursor?: string;
}

export interface LogStats {
  total: number;
  byType: {
    connection: number;
    command: number;
    telemetry: number;
    error: number;
    warning: number;
  };
}

export interface LogFilters {
  eventType?: LogEventType;
  startTime?: Date;
  endTime?: Date;
}

export const LOG_EVENT_TYPES: Record<LogEventType, { label: string }> = {
  connection: { label: "Connection" },
  command: { label: "Command" },
  telemetry: { label: "Telemetry" },
  error: { label: "Error" },
  warning: { label: "Warning" },
  info: { label: "Info" },
  update: { label: "Update" },
  registration: { label: "Registration" },
  deregistration: { label: "Deregistration" },
};

export function getEventTypeLabel(type: LogEventType): string {
  return LOG_EVENT_TYPES[type].label;
}
